aptos-move/aptos-aggregator/src/aggregator_extension.rs

```diff
@@ -218,13 +272,29 @@ impl Aggregator {
                 //     1. X >= Y: then the result is +(X-Y)
                 //     2. X  < Y: then the result is -(Y-X)
                 if self.value >= value {
+                    // This case doesn't result in an underflow.
+                    // So we are not calling record_underflow here.
                     self.value = subtraction(self.value, value)?;
                 } else {
                     // Check that we can subtract in general: we don't want to
                     // allow -10000 when limit is 10.
                     // TODO: maybe `subtraction` should also know about the limit?
-                    subtraction(self.limit, value)?;
-
+                    subtraction(self.limit, value).map_err(|err| {
+                        // TODO: The underflow value may not be correct here.
+                        self.record_underflow(value);
+                        err
+                    })?;
+
+                    if value - self.value > self.limit {
+                        self.record_underflow(value - self.value);
+                        return Err(abort_error(
+                            format!(
+                                "underflow occurred when subtracting {} from {}",
+                                value, self.value
+                            ),
+                            ESUB_UNDERFLOW,
+                        ));
+                    }
```

Comment:
let's expect on subtraction one line below that it cannot fail since self.value < value

Reply:
Done.